### PR TITLE
EditorServiceHost: allow Tcp/NamedPipe/Stdio listeners

### PR DESCRIPTION
**Pull request by yatli**
Refactored listeners and channels to be a bit more generic. The auxiliary scripts are also updated to support for example, Stdio server from outside.

### CODE REVIEW

**Comment by yatli** [comment]
oops, looks like some tests are failing? looking into it.

**Comment by yatli** [comment]
AppVeyor timeout?

**Comment by tylerl0706** [comment]
Thanks for this! Would you mind supplying the steps you used to verify/test this?

I'm interested in the "Stdio server from outside" part.

**Comment by yatli** [comment]
@tylerl0706 check this out: https://github.com/v-yadli/vimrc/blob/nvim/init.vim

here are the steps with my vimrc noise cancelled:

1. Install `neovim`
2. Use a plugin system (recommended) to install: `junegunn/fzf.vim`, `autozimu/LanguageClient-neovim`
3. Build PSES locally at `~\git\PowerShellEditorServices`
4. Add the server startup command to the neovim language service client like this: ```
"LanguageServerProtocol setup
"required for operations modifying multiple buffers like rename.
set hidden
"setup startup sequences...
let g:LanguageClient_serverCommands = {
    \ 'haskell': ['hie', '--lsp'],
    \ 'python': ['pyls'],
    \ 'cs': ['~\.omnisharp\OmniSharp.exe'],
    \ 'ps1': ['powershell', '~\git\PowerShellEditorServices\module\Start-EditorServices.ps1', '-HostName', 'nvim', '-HostProfileId', '0', '-HostVersion', '1.0.0', '-EditorServicesVersion', '1.6.0', '-LogPath', 'pses.log.txt', '-LogLevel', 'Normal', '-BundledModulesPath', '~\git\PowerShellEditorServices\module', '-Stdio'],
\ }
```
5. Now fire up a new instance of neovim and try editing a powershell script. use Ex command `:call LanguageClient_textDocument_references()` on a symbol to trigger reference finding, which will bring up a fzf prompt if you've got it installed. It's okay if fzf is not around -- the standard quickfix will be used instead.

**Comment by yatli** [comment]
This is only partially working at the moment, from pses log I see some null reference exceptions while trying to do completion, and handling document change events. Still looking into it but I think that's another issue.

**Comment by yatli** [comment]
Fixed it. Not a lot of changes so I guess we can merge it altogether?
Now it works happily with vim :)

**Comment by tylerl0706** [comment]
@daviwil can you look at this when you get a chance? I remember you saying that you had comments.

**Comment by tylerl0706** [comment]
Oh and FWIW, I was able to get NeoVim working with PSES using this PR. 🎉 🎉

**Comment by nanoant** [comment]
@yatli @tylerl0706 Just wanted to say that this PR makes PSES work like a charm with Sublime Text 3 LSP plugin  https://github.com/tomv564/LSP 🎉 😄 
powershelleditorservices-st3-lsp

I have used following `LSP.sublime-settings` (just replicated from VIM solution above).
~~~json
{
  "clients":
  {
    "poshls":
    {
      "command": [
        "powershell",
        "~\\Git\\PowerShellEditorServices\\module\\Start-EditorServices.ps1",
        "-HostName", "nvim",
        "-HostProfileId", "0",
        "-HostVersion", "1.0.0",
        "-EditorServicesVersion", "1.6.0",
        "-LogPath", "~\\Git\\PowerShellEditorServices\\pses.log.txt",
        "-LogLevel", "Normal",
        "-BundledModulesPath", "~\\Git\\PowerShellEditorServices\\module",
        "-Stdio"
      ],
      "scopes": ["source.powershell"],
      "syntaxes": ["Packages/PowerShell/Support/PowershellSyntax.tmLanguage"],
      "languageId": "powershell"
    }
  }
}
~~~

File: src/PowerShellEditorServices/Workspace/FileChange.cs [hunk 1]
```diff
@@ -34,5 +34,12 @@ public class FileChange
         /// The 1-based column offset where the change ends.
         /// </summary>
         public int EndOffset { get; set; }
+
+        /// <summary>
+        /// Indicates that the InsertString is an overwrite
+        /// of the content, and all stale content and metadata
+        /// should be discarded.
+        /// </summary>
+        public bool Reloaded { get; set; }
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Workspace/FileChange.cs]: This one is really nit picky, but could we change this member to `IsReload`

**Comment by tylerl0706** [comment]
Everything looks great! I'll give it a really good look later today but wanted to ask if you could fix this PSModulePath casing as a part of your PR:

https://github.com/PowerShell/PowerShellEditorServices/pull/629/files#diff-12fdf91ec62a52be37451ffb5af7ffb4R138

Totally optional but figured I'd ask since you're already touching the file.

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 [hunk 2]
```diff
@@ -89,12 +94,39 @@ function Start-EditorServicesHost {
...
+
+    if ($LanguageServiceNamedPipe) {
+        $languageServiceConfig.TransportType = [Microsoft.PowerShell.EditorServices.Host.EditorServiceTransportType]::NamedPipe
+        $languageServiceConfig.Endpoint = "$LanguageServiceNamedPipe"
+    }
+
+    if ($DebugServiceNamedPipe) {
+        $debugServiceConfig.TransportType = [Microsoft.PowerShell.EditorServices.Host.EditorServiceTransportType]::NamedPipe
+        $debugServiceConfig.Endpoint = "$DebugServiceNamedPipe"
+    }
```
[the remark this answers is not in the document]
Comment (tylerl0706) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: ooo that's nice 😍

**Comment by tylerl0706** [comment]
@yatli just curious, did you have an example that uses named pipes? 😄

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 [hunk 3]
```diff
@@ -31,16 +31,21 @@ function Start-EditorServicesHost {
...
         [int]
         $LanguageServicePort,
 
-        [Parameter(Mandatory=$true)]
-        [ValidateNotNullOrEmpty()]
         [int]
         $DebugServicePort,
 
+        [bool]
+        $Stdio,
```
[the remark this answers is not in the document]
Comment (rkeithhill) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: Also, might be better to change name to `UseStdio`.

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkeithhill) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: Agree with @SeeminglyScience but if you'd prefer we can add the parameter sets after this gets merged.

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 [hunk 4]
```diff
@@ -89,12 +94,39 @@ function Start-EditorServicesHost {
...
+    }
+
+    if ($DebugServiceNamedPipe) {
+        $debugServiceConfig.TransportType = [Microsoft.PowerShell.EditorServices.Host.EditorServiceTransportType]::NamedPipe
+        $debugServiceConfig.Endpoint = "$DebugServiceNamedPipe"
+    }
+
+    if ($DebugServiceOnly.IsPresent) {
+        $editorServicesHost.StartDebugService($debugServiceConfig, $profilePaths, $false);
+    } else {
```
Comment (rkeithhill) [inline module/PowerShellEditorServices/PowerShellEditorServices.psm1]: Not a huge thing but the script style for this file doesn't use cuddled else. Again, something we can tweak after merging.

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (yatli) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: ah this is totally cool!

[the remark this answers is not in the document]
Comment (yatli) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: I just thought about this for a while.. A few days ago I was debugging a problem when both language service and debug service are talking to stdio and the channel gets too crowded. In fact these two should never sit on the same stdio train. :)

**Comment by yatli** [comment]
> @yatli just curious, did you have an example that uses named pipes? 😄
@tylerl0706 nak, I just tested it working with a dummy client, with captured payload from another run :D

**Comment by yatli** [comment]
> Everything looks great! I'll give it a really good look later today but wanted to ask if you could fix this PSModulePath casing as a part of your PR:
> https://github.com/PowerShell/PowerShellEditorServices/pull/629/files#diff-12fdf91ec62a52be37451ffb5af7ffb4R138
> Totally optional but figured I'd ask since you're already touching the file.

Digging gitter chat history, I remember we discussed something about it before

**Comment by yatli** [comment]
ok, got it, `$env:PsModulePath` should be camel case.

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (yatli) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: @SeeminglyScience I tried to make it a switch as the outer starting script does, but I don't know how to relay the switch in without splitting the command into two, one with the switch and one without. Sorry I'm still learning my way on powershell 😄

**Comment by yatli** [comment]
Thank you guys for your warm and awesome input!

File: module/PowerShellEditorServices/PowerShellEditorServices.psm1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (rkeithhill) [reply module/PowerShellEditorServices/PowerShellEditorServices.psm1]: Typically it is `-InnerSwitch:$OuterSwitch`

File: module/Start-EditorServices.ps1 [hunk 5]
```diff
@@ -123,7 +135,7 @@ function Get-AvailablePort {
 
 # Add BundledModulesPath to $env:PSModulePath
 if ($BundledModulesPath) {
-    $env:PSMODULEPATH = $BundledModulesPath + [System.IO.Path]::PathSeparator + $env:PSMODULEPATH
+    $env:PsModulePath = $BundledModulesPath + [System.IO.Path]::PathSeparator + $env:PsModulePath
```
Comment (rkeithhill) [inline module/Start-EditorServices.ps1]: Close - the `s` is capitalized.  Turns out this is important on *nix platforms e.g.:
```
PSModulePath                   /home/hillr/.local/share/powershell/Modules:/usr/local/share/powershell/Modules:/opt/...
```

Comment (tylerl0706) [reply module/Start-EditorServices.ps1]: I think this is the only comment left to be addressed, (sorry for adding this on but I really appreciate it @yatli!)

Fix this and this is totally ready to be merged.

Optional: make sure vscode still works 😄

Comment (yatli) [reply module/Start-EditorServices.ps1]: @tylerl0706 I'm in a deep profiling session right now so I just quickly edited the file with Github online editor. Will be able to do heavier stuff when I get back home :)

**Comment by nanoant** [comment]
I believe I found a little problem with `Start-EditorServices.ps1`, essientially when the Stdio parent process is gone (pipe closed) `Start-EditorServices.ps1` is not exiting, but keeps working, even in the log file I get this:

   MessageReader's input stream ended unexpectedly, terminating.

As a result I end-up with stray processes. Thoughts what may cause such a behavior?

**Comment by yatli** [comment]
@nanoant The startup script simply loads the main assembly into the powershell process, which then spawns listeners and listen to incoming "connections".

Each "connection" could be a Tcp connection, a Stdio pipe, or a named pipe.
When a connection comes in, the server understands, and spawn a handler.
When the pipe is broken, the handler dies, but the listener remains.
We have discussed about this on gitter a little bit -- other implementations simply exits when the pipe is broken but it looks like the PSES listener is more like a multiplexed solution so that you can start one server and handle multiple clients. I think we would have to think a bit more about this problem, not just for Stdio.

**Comment by yatli** [comment]
But there is indeed something special about a stdio pipe -- that's a single burn medium that cannot be recovered once it's broken.

**Comment by SeeminglyScience** [comment]
This comes down to how we implement and use a custom `PSHost`.  When the language server is initialized, a single command (`Start-EditorServicesHost`) is invoked.  That command creates a new runspace with configured to use the custom host and starts the waiting for requests. At the end of that command, it awaits a task whose completion signals that the process should close.

Right now, the only time that task is completed is when

1. The language server is started as a temporary console for debugging and the debugging session has ended.

1. When an unhandled exception occurs in the background message loop.

There needs to be a new request that when received aborts command execution, disposes the `PowerShellContext` and completes that task.

[Here's the method it waits on](https://github.com/PowerShell/PowerShellEditorServices/blob/80755157599b620c8e24517e4d36166bd2943922/src/PowerShellEditorServices.Host/EditorServicesHost.cs#L338) and [here's an example of the task being set](https://github.com/PowerShell/PowerShellEditorServices/blob/80755157599b620c8e24517e4d36166bd2943922/src/PowerShellEditorServices.Host/EditorServicesHost.cs#L315).

Don't know if that necessarily needs to be in this PR, but that's how it could be done.

**Comment by nanoant** [comment]
> There needs to be a new request that when received aborts command execution, disposes the PowerShellContext and completes that task.

But when the pipe is closed there is no way anymore to connect it back or send to it anymore. This may work for TCP / named pipes but not for Stdio, moreover once the parent process that spawned PSES dies, there's no one to send such a request anymore.

Therefore I believe for Stdio we need to make it special case and shut the server down completely. Otherwise once this PR gets released there will be complaints about stray processes.

**Comment by SeeminglyScience** [comment]
> But when the pipe is closed there is no way anymore to connect it back or send to it anymore. This may work for TCP / named pipes but not for Stdio, moreover once the parent process that spawned PSES dies, there's no one to send such a request anymore.

That's a fair point.  When we detect that the stream has ended we should call `SetException` on the `TaskCompletionSource` that is holding up the process.  Probably [here](https://github.com/PowerShell/PowerShellEditorServices/blob/80755157599b620c8e24517e4d36166bd2943922/src/PowerShellEditorServices.Protocol/MessageProtocol/MessageReader.cs#L162).

I still think the the graceful close request should be implemented in cases where the editor is requesting the language server to be closed, but we should handle unexpected failures as well.

**Comment by rkeithhill** [comment]
FYI this:

> The language server is started as a temporary console for debugging and the debugging session has ended.

Changed recently to not automatically exit when the debugging session completes.  Folks, including myself, didn't like not being able to peruse over script output after the script had run.  Now what happens is that the next invocation of a temp PSIC will kill the previous one.

**Comment by tylerl0706** [comment]
@nanoant does the Sublime lsp client allow for sending/receiving custom messages?

**Comment by nanoant** [comment]
> @tylerl0706: does the Sublime lsp client allow for sending/receiving custom messages?

Well, not directly. But it is possible to run LSP module in the Sublime Python's console, e.g.:
~~~
>>> view
<sublime.View object at 0x0000015DD743B6D8>
>>> import LSP.plugin.core.clients as lsp
>>> lsp
<module 'LSP.plugin.core.clients' from 'SublimeText\\Data\\Packages\\LSP\\plugin\\core\\clients.py'>
>>> lsp.client_for_view(view)
<LSP.plugin.core.rpc.Client object at 0x0000015DD6189DA0>
>>> import LSP.plugin.core.protocol as lspproto
>>> lsp.client_for_view(view).send_request(lspproto.Request('evaluate', {'expression': '1'}), handler=lambda res: print(res))
{'variablesReference': 0, 'result': ''}
~~~

But for some reason I am unable to get any reasonable output, regardless what `'expression'` I put there, `'result'` is always empty.

**Comment by tylerl0706** [comment]
@nanoant yeah evaluate doesn't return what you'd expect. I should probably find out from @daviwil why that is.

What you want to listen for on the client side "output". For example, using vim:
```vim
" this just prints out the result
function! PS1OutputHandle(output) abort
        echomsg json_encode(a:output)
endfunction

" used as the callback for calling "evaluate". Doesn't return anything useful
function s:Log(result, error)
        :echo a:result
endfunction

" when we get a ps1, run this function
autocmd FileType ps1 call VsimEnableLanguageServerKeys()
function! VsimEnableLanguageServerKeys()

        " Note the registration of a listener that handles the output
        call LanguageClient#registerHandlers({'output': 'PS1OutputHandle'})

        " sets up keybinding
        vnoremap <silent> <F8> :call RunCode()<CR>
endfunction

function RunCode()
        " gets currently selected text and sends it with "evaluate"
        let codeString = s:get_visual_selection()
        :call LanguageClient#Call("evaluate", { 'expression': s:get_visual_selection() }, function("s:Log"))
endfunction
```

Here's a gif:
f8works

**Comment by SeeminglyScience** [comment]
I could be wrong but my assumption is that it doesn't return anything because it would have to wait for the command to finish. If it did that, I believe it would hold up the message queue, making things like the `psEditor` API impossible.

**Comment by yatli** [comment]
@tylerl0706 I guess it is like calling Process.Start() and then Process.BeginOutputReadLine(), and let the results come to the client asynchronously via custom messages.

**Comment by tylerl0706** [comment]
@yatli do you think you'll be able to finish up this PR? A lot of people are looking for more text editor options and this is the thing that enables that!

**Comment by yatli** [comment]
hey @tylerl0706 what's the missing piece?

**Comment by yatli** [comment]
@tylerl0706 ah yes, I'm also having crashes. Still haven't got time to look into that, it's getting very busy here..

**Comment by tylerl0706** [comment]
I updated to the latest neovim-language-client and intellisense is working again :)

**Comment by yatli** [comment]
oh really? trying that trick out now :)

**Comment by tylerl0706** [comment]
😍
image

**Comment by tylerl0706** [comment]
that said, we should still fix the test failures before merging.

We also should verify that this PSES still works in VSCode (it should, but just in case), but I can test that for you 🙂

**Comment by yatli** [comment]
understood ^ ^
btw nice term!

**Comment by yatli** [comment]
I think I broke this one:

```C#
            JObject result = JObject.Parse(sessionDetailsText);
            if (result["status"].Value<string>() == "started")
            {
                return new Tuple<int, int>(
                    result["languageServicePort"].Value<int>(),
                    result["debugServicePort"].Value<int>());
            }
```

but now, the details carry my modified info like this:
```json
{"status":"started","debugServiceTransport":"Tcp port 28157","languageServiceTransport":"Tcp port 14630"}
```

**Comment by tylerl0706** [comment]
Ah I see...

would a structure like this make sense:

```json
{
    "status":"started",
    "debugServiceTransport":"Tcp",
    "languageServiceTransport":"Tcp",
    "languageServicePort":1234,
    "debugServicePort":1234
}
```

where the ports are "optional" basically

**Comment by yatli** [comment]
I'm on this. A few seconds...

**Comment by yatli** [comment]
@tylerl0706 they don't go away after testing:
image

**Comment by tylerl0706** [comment]
oh... my...

**Comment by tylerl0706** [comment]
so I think this might be something that needs to be handled client side... because it's really the client that's spinning up the instance of PowerShell that runs `Start-EditorServices.ps1`

in the case of NeoVim, adding something like this on exit?

```
augroup autocom
    autocmd!
    "executes the command on quit
     autocmd VimLeave *.cpp !your_command
augroup END
```

which I got from [this](https://superuser.com/questions/653567/execute-a-command-when-vim-exits).

**Comment by tylerl0706** [comment]
I still wish we could make PSES start enough to realize the client has disconnected and stop itself but I can't think of a good way to do that at the moment.

**Comment by yatli** [comment]
@tylerl0706 it passes the tests locally now. let's wait and see. :)

**Comment by tylerl0706** [comment]
No worries, @yatli! That's awesome 😄 

Do you have any thoughts on how to address the extra processes?

**Comment by nanoant** [comment]
> I think this is mostly a problem of the language client plugin and it should be addressed over there. :)

@yatli sorry to differ, but current behavior is simply far what UNIX pipe programs are. I know I know "posh" comes from Windows world, but imagine you would need to send `shutdown` to `sed` otherwise it will be there forever with closed communication channel. This is just unnecessarily cumbersome. Name single stdin (pipe) processing program that lives when the pipe is closed.

When client dies or closes the pipe, or whatever happens with the channel, there is NO way to send shutdown command anymore, since it is single pipe, once closed it is gone forever. It is not a TCP you can reconnect etc. etc. So regardless if this will be addressed in VIM, expect stray processed anyway.

I think this should be addressed either as a part of this patch or soon after as it is strongly related.

**Comment by yatli** [comment]
@nanoant I completely agree with you on the part that `stdio` is a single-burn thing. My concern is that PSES is by-design multiplexed according to the owners -- it may span across multiple sessions, plus that it has multiple channels. What happens if the debug channel is using stdio, when it's broken, and the lsp channel is actually healthy on a Tcp port? 

Also, I'm very curious about how this thing works in VSCode -- because it looks like VSCode is connecting to PSES by two tcp ports, and how is PSES shut down when VSCode decides to do so? -- that's how I come to the conclusion that, `vim` shall shutdown `PSES` in the same way that `VSCode` shall shutdown `PSES`.

So, like I said earlier in this thread, I do think we need to think about the termination problem -- it's not just about stdio, and all clients should expect a common practice for doing so.

btw in my current setting, my vim lsp client totally ignores PSES debug port. So there'd be a thread waiting for a connection in their house.
How is PSES started? STA or MTA?

**Comment by tylerl0706** [comment]
Ok everyone, I did some research and here's what I've found:

**Right now, the client kills the server process.**

The proof can be found here in the VSCode client: https://github.com/PowerShell/vscode-powershell/blob/master/src/process.ts#L145-L161

That said, **this is not a good thing**.

You might be wondering... The [LSP has both a `shutdown` and an `exit`message](https://microsoft.github.io/language-server-protocol/specification#shutdown), don't we implement those?

The answer is yes.......... but it's functionally a no-op 😢

You can see the `TODO`comment here:
https://github.com/PowerShell/PowerShellEditorServices/blob/master/src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs#L148

which is referring to this TaskCompletionSource that is causing the process to stay open:
https://github.com/PowerShell/PowerShellEditorServices/blob/4716b7f4f71ec3a6645720e6b427c2dba21c89ba/src/PowerShellEditorServices.Host/EditorServicesHost.cs#L341

**So what's the plan?**

Since we currently rely on the client to kill the server, we should call that "out of scope" of this PR.

I'll open up a new issue with priority (that I will address ASAP) that actually implements these messages correctly.

**Comment by tylerl0706** [comment]
@nanoant + @yatli does that seem fair?

**Comment by tylerl0706** [comment]
P.S. I verified that the latest changes work with VSCode 🎉

**Comment by rjmholt** [comment]
I've had a look through everything and this all looks really cool. Haven't had a chance to run the code yet, but will try to this weekend.

On the PSES shutdown thing: is there a way for us to keep track of how many clients we are serving and then shutdown when it goes to zero (like a Windows named pipe)?

**Comment by tylerl0706** [comment]
> how many clients

Just for information's sake, this should only be 1 or 0. 2 clients cannot share an instance of PSES because a single runspace is being managed in each PSES session.

> shutdown when it goes to zero

I believe that's only possible by using a heartbeat 😟 I'm not a huge fan of that idea, but even if we wanted to do that, I'd still call that out of scope of this PR IMO.

**Comment by rjmholt** [comment]
> I'd still call that out of scope of this PR IMO.

Definitely out of scope for this PR, yes.

> this should only be 1 or 0. 2 clients cannot share an instance of PSES because a single runspace is being managed in each PSES session.

So does that mean that for a single PSES process there can only be one client? So the problem is that a client connects to PSES over multiple streams and we don't have a good answer for what to do when those streams break?

My question is, do all our proposed streams guarantee connection? I know other protocols have heartbeat messages, but the concept of a connection-oriented transport layer in my mind means we can detect when a connection is broken or closed. When all connections to PSES are closed, we should just wrap up. Generally the client should give that command, but they can do it with a `shutdown` message. I'm assuming there's a complicating factor here that I've missed though.

**Comment by tylerl0706** [comment]
I opened this issue to chat more about that:

https://github.com/PowerShell/PowerShellEditorServices/issues/655

@nanoant, is [this](https://github.com/PowerShell/PowerShellEditorServices/pull/629#issuecomment-383231416) cool with you?

I'm going to accept the PR on Monday if no one has any concerns.

**Comment by nanoant** [comment]
@tylerl0706 Thanks. I am okay to continue the shutdown discussion at #655 and have this merged.

**Comment by tylerl0706** [comment]
It's merged!! 🎉 Thank you so much @yatli!

Also thank you so much to @nanoant for all the awesome feedback. The release of this will come this week and it would be really cool to see a Sublime and NeoVim PSES guide after that 😁

**Comment by tylerl0706** [comment]
PSES 1.7.0 is released with this change!

https://github.com/PowerShell/PowerShellEditorServices/releases/tag/v1.7.0

You should be able to download this zip and get it to work with your LSP client of choice

**Pull request merged**